File: ios/Classes/EventDispatcher.swift
import Flutter
import Foundation

/// Thread-safe bridge to a Flutter `FlutterEventChannel`. Frame analysis runs on
/// a background queue, but the event sink must be invoked on the main thread —
/// this marshals every emission to main and tolerates the absence of a listener.
final class EventDispatcher: NSObject, FlutterStreamHandler {
  private var sink: FlutterEventSink?

  func onListen(
    withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink
  ) -> FlutterError? {
    sink = events
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    sink = nil
    return nil
  }

  func send(_ payload: [String: Any?]) {
    guard let sink = sink else { return }
    if Thread.isMainThread {
      sink(payload)
    } else {
      DispatchQueue.main.async { [weak self] in self?.sink?(payload) }
    }
  }

  func sendError(_ code: String, _ message: String?) {
    send(["type": EventType.error, "code": code, "message": message as Any?])
  }

  func dispose() { sink = nil }
}

File: android/src/main/kotlin/com/umda/barcode_scanner_pro/EventDispatcher.kt
package com.umda.barcode_scanner_pro

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel

/**
 * Thread-safe bridge to a Flutter [EventChannel]. Frames are analyzed on a
 * background executor, but [EventChannel.EventSink] must be invoked on the main
 * thread — this class marshals every emission via the main looper and tolerates
 * the sink being absent (no listener yet) without throwing.
 */
internal class EventDispatcher : EventChannel.StreamHandler {
    private val mainHandler = Handler(Looper.getMainLooper())
    @Volatile private var sink: EventChannel.EventSink? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }

    fun send(payload: Map<String, Any?>) {
        val s = sink ?: return
        if (Looper.myLooper() == Looper.getMainLooper()) {
            s.success(payload)
        } else {
            mainHandler.post { sink?.success(payload) }
        }
    }

    fun sendError(code: String, message: String?) {
        send(mapOf("type" to EventType.ERROR, "code" to code, "message" to message))
    }

    fun dispose() {
        sink = null
        mainHandler.removeCallbacksAndMessages(null)
    }
}
